pkg/ops/ops.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2016 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ops
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/appc/cni/pkg/ns"
+
+	"github.com/vishvananda/netlink"
+)
+
+type NetOps interface {
+	LinkAdd(link netlink.Link) error
+	LinkDel(link netlink.Link) error
+	LinkByIndex(index int) (netlink.Link, error)
+	LinkByName(name string) (netlink.Link, error)
+	LinkSetName(link netlink.Link, name string) error
+	LinkSetUp(link netlink.Link) error
+	LinkSetMaster(link netlink.Link, master *netlink.Bridge) error
+	LinkSetNsFd(link netlink.Link, fd int) error
+	LinkList() ([]netlink.Link, error)
+
+	AddrAdd(link netlink.Link, addr *netlink.Addr) error
+	AddrList(link netlink.Link, family int) ([]netlink.Addr, error)
+
+	GetNS(nspath string) (ns.NetNS, error)
+	NewNS(persist, switchBack bool) (ns.NetNS, error)
+	WithNetNSPath(nspath string, lockThread bool, f func(ns.NetNS) error) error
+}
+
+var instance NetOps
+var mutex sync.Mutex
+
+func newNetOps(testOps bool) NetOps {
```

Comment:
I guess you could do something like:

```
func A() {
    netops := GetNetOps()
    netops.LinkAdd()
    netops.LinkList()
    B(<stuff>)
}

func B() {
    netops := GetNetOps()
    netops.Whatever().
}
```

but that's still basically using a singleton because we need to create the test netops object once at the beginning and then return that from GetNetOps().  And we still have to clear it unless we pass the same object around to every function.